File: src/main/java/io/klbz/curie/FirstFocus.java
package io.klbz.curie;

import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;

public class FirstFocus<F, S> {
	private final Pair<F, S> pair;

	private FirstFocus(Pair<F, S> pair){ this.pair = pair; }

	public static <F, S> FirstFocus<F, S> on(Pair<F, S> pair){ return new FirstFocus<>(pair); }

	public Pair<F, S> unfocus(){ return pair; }

	public <FPrime> FirstFocus<FPrime, S> replace(FPrime newFirst){ return on(pair.replaceFirst(newFirst)); }

	public F isolate(){ return pair.isolateFirst(); }

	public <FPrime> FirstFocus<FPrime, S> map(Function<? super F, FPrime> f){ return on(pair.mapFirst(f)); }

	public FirstFocus<F, S> withDo(Consumer<F> doF){
		pair.withFirstDo(doF);
		return this;
	}

	@Override
	public boolean equals(Object o){
		if(this == o) return true;
		if(o == null) return false;

		if(o.getClass() == Pair.class) return Objects.equals(pair, o);

		if(o.getClass() == FirstFocus.class){
			FirstFocus<?, ?> that = (FirstFocus<?, ?>) o;
			return Objects.equals(pair, that.pair);
		}

		if(o.getClass() == SecondFocus.class){
			SecondFocus<?, ?> that = (SecondFocus<?, ?>) o;
			return Objects.equals(pair, that.unfocus());
		}

		return false;
	}

	@Override
	public int hashCode(){ return pair.hashCode(); }

	@Override
	public String toString(){ return "FirstFocus{" + pair + '}'; }
}

File: src/main/java/io/klbz/curie/FirstFocus.kt
package io.klbz.curie

import java.util.function.Consumer
import java.util.function.Function

class FirstFocus<F, S> internal constructor(private val pair: Pair<F, S>) {

    fun unfocus() = pair

    fun <FPrime> replace(newFirst: FPrime) = on(pair.replaceFirst(newFirst))

    fun isolate() = pair.isolateFirst()

    fun <FPrime> map(f: Function<in F, FPrime>) = on(pair.mapFirst(f))

    fun withDo(doF: Consumer<F>) = also { pair.withFirstDo(doF) }

    //region Kotlin-specific overloads

    @JvmSynthetic
    fun <FPrime> map(f: (F) -> FPrime) = on(pair.mapFirst(f))

    @JvmSynthetic
    fun withDo(doF: (F) -> kotlin.Unit) = also { pair.withFirstDo(doF) }

    //endregion

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false

        if (other.javaClass == Pair::class.java) return pair == other

        if (other.javaClass == FirstFocus::class.java) {
            val that = other as FirstFocus<*, *>?
            return pair == that!!.pair
        }

        if (other.javaClass == SecondFocus::class.java) {
            val that = other as SecondFocus<*, *>?
            return pair == that!!.unfocus()
        }

        return false
    }

    override fun hashCode() = pair.hashCode()

    override fun toString() = "FirstFocus{$pair}"

    companion object {
        @JvmStatic
        fun <F, S> on(pair: Pair<F, S>): FirstFocus<F, S> {
            return FirstFocus(pair)
        }
    }
}
